sycl/include/CL/sycl/types.hpp

```diff
@@ -280,7 +309,154 @@ using Rtp = detail::bool_constant<Mode == rounding_mode::rtp>;
+
+#undef __SYCL_GENERATE_CONVERT_IMPL
+
+// unsigned to unsigned
+#define __SYCL_GENERATE_CONVERT_IMPL(DestType)                                 \
+  template <typename T, typename R, rounding_mode roundingMode>                \
+  detail::enable_if_t<                                                         \
+      !std::is_same<T, R>::value && is_uint_to_uint<T, R>::value &&            \
+          std::is_same<R, DestType>::value &&                                  \
+          std::is_same<cl::sycl::detail::ConvertToOpenCLType_t<T>, R>::value,  \
+      R>                                                                       \
+  convertImpl(T Value) {                                                       \
+    return static_cast<R>(Value);                                              \
+  }
+
+__SYCL_GENERATE_CONVERT_IMPL(uchar)
+__SYCL_GENERATE_CONVERT_IMPL(ushort)
+__SYCL_GENERATE_CONVERT_IMPL(uint)
+__SYCL_GENERATE_CONVERT_IMPL(ulong)
+
+#undef __SYCL_GENERATE_CONVERT_IMPL
+
+#define __SYCL_GENERATE_CONVERT_IMPL(DestType)                                 \
+  template <typename T, typename R, rounding_mode roundingMode>                \
+  detail::enable_if_t<                                                         \
+      !std::is_same<T, R>::value && is_uint_to_uint<T, R>::value &&            \
+          std::is_same<R, DestType>::value &&                                  \
+          !std::is_same<cl::sycl::detail::ConvertToOpenCLType_t<T>, R>::value, \
+      R>                                                                       \
+  convertImpl(T Value) {                                                       \
+    using OpenCLT = cl::sycl::detail::ConvertToOpenCLType_t<T>;                \
+    OpenCLT OpValue = cl::sycl::detail::convertDataToType<T, OpenCLT>(Value);  \
+    return __spirv_UConvert##_R##DestType(OpValue);                            \
+  }
+
+__SYCL_GENERATE_CONVERT_IMPL(uchar)
+__SYCL_GENERATE_CONVERT_IMPL(ushort)
+__SYCL_GENERATE_CONVERT_IMPL(uint)
+__SYCL_GENERATE_CONVERT_IMPL(ulong)
+
+#undef __SYCL_GENERATE_CONVERT_IMPL
+
+// unsigned to (from) signed
+template <typename T, typename R, rounding_mode roundingMode>
+detail::enable_if_t<is_sint_to_from_uint<T, R>::value, R> convertImpl(T Value) {
+  return static_cast<R>(Value);
+}
+
+// sint to float
+#define __SYCL_GENERATE_CONVERT_IMPL(SPIRVOp, DestType)                        \
+  template <typename T, typename R, rounding_mode roundingMode>                \
+  detail::enable_if_t<                                                         \
+      is_sint_to_float<T, R>::value && std::is_same<R, DestType>::value, R>    \
+  convertImpl(T Value) {                                                       \
+    using OpenCLT = cl::sycl::detail::ConvertToOpenCLType_t<T>;                \
+    OpenCLT OpValue = cl::sycl::detail::convertDataToType<T, OpenCLT>(Value);  \
+    return __spirv_Convert##SPIRVOp##_R##DestType(OpValue);                    \
+  }
```

Comment:
This `std::is_same<R, DestType::value` doesn't seem like a right thing to do here. Is it possible to avoid adding this check into each `enable_if`?

What about creating partial specializations of `convertImpl`?

```c++
template <typename T, typename R, rounding_mode roundingMode>
R convertImp(T Value) { 
  // the most generic one
  // static_cast as fallback?
}

template <typename T, rounding_mode roundingMode>
half convertImpl<T, half, roundingMode>(half Value) {
  // actual implementation here
}

// Actual implementations can still be generated by preprocessor:
#define __SYCL_GENERATE_CONVERT_IMPL(SPIRVOp, DestType) \
detail::enable_if_t< is_signed_arithmetic<T>::value, T> \
template <typename T, rounding_mode roundingMode> \
convertImpl<T, DestType, roundingMode>(DestType Value) { \
  // actual implementation here \
}

__SYCL_GENERATE_CONVERT_IMPL(SToF, half)
__SYCL_GENERATE_CONVERT_IMPL(SToF, float)
__SYCL_GENERATE_CONVERT_IMPL(SToF, double)

#undef __SYCL_GENERATE_CONVERT_IMPL
```

Reply:
How about just get name of a type as string and add it to a spirv func name?

Reply:
I don't understand how to implement your idea

Reply:
C++ doesn't support function template partial specialization.